homeassistant/components/tplink/light.py

```diff
@@ -72,192 +75,327 @@ def brightness_from_percentage(percent):
-            self.hass.add_job(self.do_update)
+            self.hass.add_job(self.do_update_retry, True)
 
-    def do_update(self):
-        """Update states."""
+    def do_update_retry(self, update_state_machine: bool) -> None:
+        """Update state data with retry.""" ""
         try:
-            if self._supported_features is None:
-                self.get_features()
-
-            self._state = self.smartbulb.state == SmartBulb.BULB_STATE_ON
-
-            if self._supported_features & SUPPORT_BRIGHTNESS:
-                self._brightness = brightness_from_percentage(self.smartbulb.brightness)
-
-            if self._supported_features & SUPPORT_COLOR_TEMP:
-                if (
-                    self.smartbulb.color_temp is not None
-                    and self.smartbulb.color_temp != 0
-                ):
-                    self._color_temp = kelvin_to_mired(self.smartbulb.color_temp)
-
-            if self._supported_features & SUPPORT_COLOR:
-                hue, sat, _ = self.smartbulb.hsv
-                self._hs = (hue, sat)
-
-            if self.smartbulb.has_emeter:
-                self._emeter_params[ATTR_CURRENT_POWER_W] = "{:.1f}".format(
-                    self.smartbulb.current_consumption()
-                )
-                daily_statistics = self.smartbulb.get_emeter_daily()
-                monthly_statistics = self.smartbulb.get_emeter_monthly()
-                try:
-                    self._emeter_params[ATTR_DAILY_ENERGY_KWH] = "{:.3f}".format(
-                        daily_statistics[int(time.strftime("%d"))]
-                    )
-                    self._emeter_params[ATTR_MONTHLY_ENERGY_KWH] = "{:.3f}".format(
-                        monthly_statistics[int(time.strftime("%m"))]
-                    )
-                except KeyError:
-                    # device returned no daily/monthly history
-                    pass
-
-            self._available = True
-
+            # Update light features only once.
+            self._light_features = (
+                self._light_features or self.get_light_features_retry()
+            )
+            self._light_state = self.get_light_state_retry(self._light_features)
+            self._is_available = True
         except (SmartDeviceException, OSError) as ex:
-            if self._available:
+            if self._is_available:
                 _LOGGER.warning(
-                    "Could not read state for %s: %s", self.smartbulb.host, ex
+                    "Could not read data for %s: %s", self.smartbulb.host, ex
                 )
-            self._available = False
+            self._is_available = False
+
+        # The local variables were updates asyncronousally,
+        # we need the entity registry to poll this object's properties for
+        # updated information. Calling schedule_update_ha_state will only
+        # cause a loop.
+        if update_state_machine:
+            self._async_write_ha_state()
```

Comment:
Just scheduling a state update won't cause a loop. There will be a loop if we force an update too.
```suggestion
            self.schedule_update_ha_state()
```

Reply:
I think the reason this is causing a loop is because the update method for this entity runs the API calls on a background task. So calling schedule_update_ha_state() results in calling update again from the main loop. Which causes a background API call and another update.

Calling _async_write_ha_state() avoids this issue by not calling Entity.async_update again. I know we're not suppose to call private method but I don't see any better way to avoid this.

Reply:
No. As I said, there will only be a loop if we call `self.schedule_update_ha_state(True)`.

Reply:
Thanks for catching that!